lib/matplotlib/axis.py

```diff
@@ -1437,32 +1435,37 @@ def grid(self, b=None, which='major', **kwargs):
             Define the line properties of the grid, e.g.::
 
                 grid(color='r', linestyle='-', linewidth=2)
-
         """
-        if len(kwargs):
-            if not b and b is not None:  # something false-like but not None
+        if b is not None:
+            if 'visible' in kwargs and bool(b) != bool(kwargs['visible']):
+                raise ValueError(
+                    "'b' and 'visible' specify inconsistent grid visibilities")
+            if kwargs and not b:  # something false-like but not None
                 cbook._warn_external('First parameter to grid() is false, '
                                      'but line properties are supplied. The '
                                      'grid will be enabled.')
-            b = True
+                b = True
         which = which.lower()
         _api.check_in_list(['major', 'minor', 'both'], which=which)
         gridkw = {'grid_' + item[0]: item[1] for item in kwargs.items()}
+        if 'grid_visible' in gridkw:
+            forced_visibility = True
+            gridkw['gridOn'] = gridkw.pop('grid_visible')
+        else:
+            forced_visibility = False
 
         if which in ['minor', 'both']:
-            if b is None:
-                self._gridOnMinor = not self._gridOnMinor
-            else:
-                self._gridOnMinor = b
-            self.set_tick_params(which='minor', gridOn=self._gridOnMinor,
-                                 **gridkw)
+            if b is None and not forced_visibility:
+                gridkw['gridOn'] = not self._minor_tick_kw['gridOn']
```

Comment:
Should this toggle the value in `_minor_tick_kw`?

Reply:
set_tick_params (called just below) does that.